text/3803-enum-repr-no-niches.md

```diff
@@ -0,0 +1,709 @@
+assert_eq!(Weather::Sunny as u8, 0);
+assert_eq!(snowy as u8, 3);
+```
+
+Note that this RFC does not propose a way to safely construct enums of the unknown variants, see [the future possibilities section][future-possibilities] for that.
+
+## Discriminant
+
+The `core::mem::Discriminant` of the enum exposed to the user should differ based on the actual underlying value, regardless of the variants not being available in the source code. Building on the previous example:
+
+```rust
+use core::mem::discriminant;
+assert_eq!(discriminant(&snowy), discriminant(&snowy));
+assert_ne!(discriminant(&snowy), discriminant(&Weather::Sunny));
+assert_ne!(discriminant(&snowy), discriminant(&Weather::Windy));
+assert_ne!(discriminant(&snowy), discriminant(&Weather::Rainy));
+assert_ne!(discriminant(&snowy), discriminant(&Weather::from(42)));
+```
+
+## Usage in pattern matching
+
+Adding the `no_niches` modifier affects exhaustiveness-checking, since it adds extra variants.
+
+Unlike `#[non_exhaustive]`, the check is enabled everywhere, regardless of visibility attributes and crate boundaries (because it works on the ABI level).
+
+Whether exhaustiveness-checking is affected by `no_niches` enums that already has enough variants to represent every bit-pattern of the underlying integer type depends on the final syntax, see [the unresolved questions section][unresolved-questions].
+
+## Derives
+
+`#[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]` compares the discriminant, regardless of the source-ordering (like they do today).
+
+`#[derive(Default)]` continues to require that a variant is marked with `#[default]`.
+
+The exact output of `#[derive(Debug)]` is [an unresolved question][derivedebug].
+
+User-defined derives and proc-macros will have to know when to emit the extra match arm (so most user-defined derive macros probably won't support `#[repr(_, no_niches)]` initially). Proc-macros could either choose to inspect the `repr`, or to defensively emit an extra `#[allow(unreachable_patterns)] _ => {}` match arm.
+
+## Handling unknown variants
+
+Matching on unknown variants is possible:
+
+```rust
+const SNOWY: Weather = unsafe { core::mem::transmute::<u8, Weather>(3) };
+
+match Weather::current() {
+    Weather::Sunny => println!("Nice an sunny!"),
+    Weather::Windy => println!("A bit breezy."),
+    Weather::Rainy => println!("Let's go inside and grab a cup of tea."),
+    SNOWY => println!("Do you wanna build a snowman?"),
+    weather => eprintln!("Unable to determine the weather, got enum value {}.", weather as u8),
+}
+```
+
+And would continue to work even if the `Weather::Snowy` variant was added in the future.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This adds surface area to the language for something that could be considered a relic from the C past.
+
+It may be confusing when to use this over `#[non_exhaustive]` (see also the syntax discussion in [the unresolved questions section][unresolved-questions]).
+
+
+# Rationale and alternatives
```

Comment:
To me this looks like a way to spell

```rust
union CWeather {
   w: Weather,
   unknown: u8
}
```

Wouldn't it also be an option to teach `match` and its discriminant-loads to soundly look through these kinds of unions so one can write

```rust
match ffi_val {
   CWeather { w: Weather::Windy } => {...}
   CWeather { w: Weather::Rainy } => {...}
   CWeather { unknown: SNOWY } => {...}
}
```

The advantage of this approach would be that one can talk about the _known-variants_ and _potentially unknown junk from C_ cases as separate types.

Reply:
This has a few disadvantages:

* you have to ensure that the type of the `unknown` field and `Weather`’s repr match
* not only does matching against union fields require `unsafe`, matching against the `w` field is [unsound](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024) if the scrutinee doesn’t have a valid bit pattern for that field

Reply:
Thanks for the suggestion, I added it to the alternatives section with a few advantages and disadvantages. Feel free to comment further there if you have suggestions for better wording.